docker_node: Module for swarm node operations

WojciechowskiPiotr opened the pull request:
##### SUMMARY
New state 'node' that allows changing Swarm node availability and role or returns node_facts about the node from Swarm manager

##### ISSUE TYPE
* Feature Pull Request
 
##### COMPONENT NAME
docker_swarm

##### ADDITIONAL INFORMATION
This is an extension to docker_swarm, but if required it could be easily a new module. The current docker modules do not support Swarm nodes role and availability changes. I use the API update_node call to make changes.

Commit:
* docker_swarm: new state 'node' which allows changing node availability and role

Commit:
* docker_swarm: Returning node_facts read from Swarm manager if no additional parameters than required 'node_id' provided for 'state: node'

Commit:
lib/ansible/modules/cloud/docker/docker_swarm.py:199:1: W293 blank line contains whitespace
lib/ansible/modules/cloud/docker/docker_swarm.py:205:1: W293 blank line contains whitespace
lib/ansible/modules/cloud/docker/docker_swarm.py:0:0: E309 version_added for new option (node_availability) should be 2.8. Currently 0.0
lib/ansible/modules/cloud/docker/docker_swarm.py:0:0: E309 version_added for new option (node_role) should be 2.8. Currently 0.0

tbouvet commented:
@WojciechowskiPiotr I think it will be better to change the state's name. `Node` is not a state, maybe you can use `update` instead.

felixfontein commented on lib/ansible/modules/cloud/docker/docker_swarm.py:
```diff
@@ -498,12 +596,15 @@ def main():
         ('state', 'join', ['advertise_addr', 'remote_addrs', 'join_token']),
-        ('state', 'remove', ['node_id'])
+        ('state', 'remove', ['node_id']),
+        ('state', 'node', ['node_id'])
```
Can you add a trailing comma, to simplify future diffs?

felixfontein commented on lib/ansible/modules/cloud/docker/docker_swarm.py:
```diff
@@ -498,12 +596,15 @@ def main():
-        rotate_manager_token=dict(type='bool', default=False)
+        rotate_manager_token=dict(type='bool', default=False),
+        node_availability=dict(type='str', choices=['active', 'pause', 'drain']),
+        node_role=dict(type='str', choices=['worker', 'manager'])
```
Can you add a trailing comma, to simplify future diffs?

felixfontein commented:
How about `update-node` instead of `update` (which is pretty generic)?

Commit:
* docker_swarm: tailing commas on last entry o dict objects

Commit:
* docker_swarm: State to update swarm node changed from 'node' to 'node-update'

Commit:
* docker_swarm: RETURN description for node_facts when running state 'node-update'

WojciechowskiPiotr commented:
@tbouvet @felixfontein updated as per your recommendations. 

Every successful call of state 'node-update' returns note_facts about the node in response. If someone wants only read the node_facts, it is required to call state 'node-update' with no additional dedicated parameters specific for this state, at this moment those are 'note_availability' and 'node_role' (might be extended by labels manipulation in future). Should I leave this approach or alight with the state 'inspect', so create new state 'node-inspect' to fetch node_facts from the swarm manager and return an error in 'node-update' when no change parameters are provided?

tbouvet commented:
@WojciechowskiPiotr I think it will be better to let `inspect` state and to add your node informations to this state. 
You can add a result like `node_facts` in the results if a node_id is present.

felixfontein commented:
For only reading facts, there should be a separate `_facts` module. Also, the `state == 'inspect'` mode should probably not have been there in the first place, but that's too late to change now :)

It probably makes sense to add a `docker_swarm_facts` module (which does what `state == 'inspect'` currently does) and to deprecate `state == 'inspect'`, and add maybe another module `docker_swarm_node_facts` which allows to retrieve facts about a node in a docker swarm.

Actually, now that I'm thinking about it, one could also move this feature into a new module `docker_swarm_node`, which allows to configure a node in a docker swarm. Such a solution might also improve the situation for #49547, since the `labels` option seems to be confusing where it applies to.

@tbouvet any opinion?

felixfontein commented:
We don't have to split anything, I just wanted to mention it because it came up to my mind ;-) The advantage of adding a module `docker_swarm_node` would probably be that both modules become easier to use as the interface is simpler for both of them (it's clearer which option is used for what). `docker_swarm` would be for managing global swarm state, and `docker_swarm_node` for managing local swarm state (for a node). I'm not using docker swarms, so I'm also not using the module, so maybe this separation doesn't make too much sense, though.

tbouvet commented:
For me to split means : create a new module `doker_swarm_facts` instead of having `inspect` state in `docker_swarm` like you suggest. So I agree with you.

You are right for the 2 modules swarm and swam_node, it will be easier to use but I think it doesn't make too much sense. Otherwise, `docker_swarm` will only have one function : **create/delete** a cluster. But a cluster without node, it's no sense.

WojciechowskiPiotr commented:
The advantage I see in splitting node operations from swarm is keeping consistency with the command line. It is `docker node` and `docker swarm` to perform operations. In such case the new modules names might be `docker_node` and `docker_node_facts` instead of proposed `docker_swarm_node`. We already have `docker_image_facts` while for `docker_container` or `docker_network` it is not clear how to just read the facts so I must agree such split will make it simpler.

That will also help to solve #49547 because labels are applied to nodes via manager not to swarm stack as itself.

To keep a more consistent approach there should be `docker_swarm_facts` which is easy to create and state `inspect` might be depreciated in further releases.

tbouvet commented:
So ok, I agree with you. Good arguments 👍

Commit:
* docker_node_facts: new module to inspect swarm node via swarm manager

Commit:
* docker_node: module for swarm node management (availability and role changes at the moment)

Commit:
* docker_swarm: revert back to orinal ansible/devel

Commit:
* docker_swarm_facts: new module to read docker swarm cluster information, will replace docker_swarm state 'inspect' in future
* docker_node_facts: update in module description

Commit:
* docker_node: #49547 docker swarm node labels manipulation

Commit:
* docker_node: #49547 docker swarm node labels manipulation

Commit:
* docker_node: #49547 docker swarm node labels manipulation documentation update

WojciechowskiPiotr commented:
As per discussion I created three modules `docker_node` for node parameters manipulation, `docker_node_facts` to inspect current node parameters and also `docker_swarm_facts` to read current swarm parameters. I also reverted back the `docker_swarm` to the version from the official repository.  

I am not sure if now I have to split this PR into three separate PRs for each module? Please advise before I either open new PRs or rebase current branch.

felixfontein commented:
I think it's best to create two new PRs (for `docker_swarm_facts` and `docker_node_facts`), and keep the `docker_node` module in this PR. That way you can edit `docker_swarm` and `docker_node` in parallel in this PR. The `*_facts` modules should be pretty straightforward anyway...

And it's great to know that you like to be involved in maintaining the modules :-) (You might also be interested in ansible/community#408)

felixfontein commented on lib/ansible/modules/cloud/docker/docker_swarm_facts.py:
```diff
@@ -0,0 +1,119 @@
+
+
+def main():
+    option_minimal_versions = dict(
```
You can remove this, it's not needed (since your module has no options). Also remove the `option_minimal_versions=option_minimal_versions,` line below in the `AnsibleDockerClient()` constructor call.

felixfontein commented on lib/ansible/modules/cloud/docker/docker_node_facts.py:
```diff
@@ -0,0 +1,127 @@
+        name=dict(type='str', required=True),
+    )
+
+    option_minimal_versions = dict(
```
Same here, this can go.

felixfontein commented on lib/ansible/modules/cloud/docker/docker_node_facts.py:
```diff
@@ -0,0 +1,127 @@
+        supports_check_mode=True,
+        min_docker_version='2.6.0',
+        min_docker_api_version='1.25',
+        option_minimal_versions=option_minimal_versions,
```
This too.

felixfontein commented on lib/ansible/modules/cloud/docker/docker_swarm_facts.py:
```diff
@@ -0,0 +1,119 @@
+
+'''
+
+RETURN = '''
```
What does it return when the current Docker daemon is not part of a swarm? (Then this module could also be used to detect whether Docker daemon is part of a swarm.)

felixfontein commented on lib/ansible/modules/cloud/docker/docker_swarm_facts.py:
```diff
@@ -0,0 +1,119 @@
+'''
+
+EXAMPLES = '''
+- name: Get info on node
```
Shouldn't this be 'swarm' (and not 'node')?

felixfontein commented on lib/ansible/modules/cloud/docker/docker_swarm_facts.py:
```diff
@@ -0,0 +1,119 @@
+        swarm_info = json.loads(json_str)
+        return swarm_info
+    except APIError:
+        return
```
```suggestion
        return None
```

felixfontein reviewed (commented):
Some comments on the facts modules.

felixfontein commented:
Thinking about it, I would suggest to still name the modules `docker_swarm_node*` instead of `docker_node*`. That groups them closer together in the module index and makes it clearer to people not using Docker swarm that these modules aren't of interest to them.

This is just my personal opinion, I'm leaving the decision to you two (@tbouvet @WojciechowskiPiotr).

tbouvet commented on lib/ansible/modules/cloud/docker/docker_node.py:
```diff
@@ -0,0 +1,338 @@
+        self.results['actions'].append("Node updated")
+        self.results['changed'] = True
+
+    def node_list(self):
```
Maybe this function could be in `docker_swarm_facts` module because it's all nodes in a swarm environment.

WojciechowskiPiotr replied:
I've checked several other `_facts` modules from the `cloud` category and all of them returns the full information of all existing objects or just one\multiple objects specified usually as a `name` parameter. But it is always the full structure.

To be more align with the docker command line I am thinking about changing the `docker_node_facts` by setting the `name` as optional and when not provided return information about all nodes, or just one if `name` is provided. Then amend the `docker_node` state `list` so data returned by method matches the command line output.

felixfontein replied:
Making `name` optional is probably a good idea.

tbouvet commented on lib/ansible/modules/cloud/docker/docker_node.py:
```diff
@@ -0,0 +1,338 @@
+
+        choice_map.get(self.parameters.state)()
+
+    def __isSwarmManager(self):
```
Same function in `docker_swarm` module. It will be better to create a new module `docker_swarm_common` to assemble common code.

felixfontein replied:
That would be a Python module in lib/ansible/module_utils, not a Ansible module (in case this confuses anyone). I guess just naming it `docker_swarm` would be fine, too (everything in module_utils is already expected to be common to multiple modules).

WojciechowskiPiotr replied:
But that will require, except changing the name of methods, provide them as argument the `self.client.info()` or `self.client.inspect_swarm()` results or the `Client` object itself to call the method inside. I think the second approach align more with Ansible coding standards

felixfontein replied:
From the way the docker modules currently operate, it probably makes sense to add a `AnsibleDockerSwarmClient` to `module_utils/docker_swarm.py`, which extends `AnsibleDockerClient`. Then `docker_swarm`, `docker_swarm_facts`, `docker_node` and `docker_node_facts` could use `AnsibleDockerSwarmClient` instead of using `AnsibleDockerClient` directly.

WojciechowskiPiotr replied:
I have no problem with this approach

WojciechowskiPiotr replied:
Implemented as new class now

tbouvet commented on lib/ansible/modules/cloud/docker/docker_node.py:
```diff
@@ -0,0 +1,338 @@
+        except APIError:
+            return False
+
+    def __isSwarmNode(self):
```
Same function in `docker_swarm` module. It will be better to create a new module `docker_swarm_common` to assemble common code.

WojciechowskiPiotr commented:
@felixfontein I would stick to `docker_node` because of the following reasons:



> Thinking about it, I would suggest to still name the modules `docker_swarm_node*` instead of `docker_node*`. That groups them closer together in the module index and makes it clearer to people not using Docker swarm that these modules aren't of interest to them.
> 
> This is just my personal opinion, I'm leaving the decision to you two (@tbouvet @WojciechowskiPiotr).

I think it depends if modules should reflect the command line and API from Docker or we want to name them independently. It is still `docker node` and `/nodes` API, the `docker service` and `/services` API, the `docker secret` and `/secrets` API... For all Swarm related commands only the `docker service` module name is `docker_swarm_service` but I believe it is because `docker_service` was already in use for a module that reflects the `docker-compose` usage (the module name is quite misleading in my opinion). 

I don't know the reason why docker developers decided to not put them under `swarm` category in a command line or API, but I can think of one - right now those commands are related to Swarm, but in future, they might be extended to support other types of orchestrators. The `docker stack` already support Kubernetes and Swarm. 

Lastly, I think most people who will use the modules first got to know the Swarm via command line so it will be more natural for them to understand `docker_node` than `docker_swarm_node`. At least it will be for me.

So unless @tbouvet disagree with me I opt for naming that align with the docker command line and API.

WojciechowskiPiotr committed:
Delete docker_node_facts.py

This file will be presented in separate PR

WojciechowskiPiotr committed:
Delete docker_swarm_facts.py

This file will be added in separate PR

Commit:
docker_node_facts: new module to read information about docker swarm node from swarm manager

Commit:
Revert "docker_node_facts: new module to read information about docker swarm node from swarm manager"

This reverts commit 133a8a0c0c41fe4b433b2b79885689a01a9f7bd1.

felixfontein commented:
> I think it depends if modules should reflect the command line and API from Docker or we want to name them independently.

That's probably a matter of taste :) I would vote to stick mostly to the command line, but extend names to make them clearer.

> For all Swarm related commands only the `docker service` module name is `docker_swarm_service` but I believe it is because `docker_service` was already in use for a module that reflects the `docker-compose` usage (the module name is quite misleading in my opinion).

I think so too, and I fully agree ;-)

> I don't know the reason why docker developers decided to not put them under `swarm` category in a command line or API, but I can think of one - right now those commands are related to Swarm, but in future, they might be extended to support other types of orchestrators. The `docker stack` already support Kubernetes and Swarm.

That's one more reason to name them more explicit in Ansible, since other orchestrators have their own modules.

> Lastly, I think most people who will use the modules first got to know the Swarm via command line so it will be more natural for them to understand `docker_node` than `docker_swarm_node`. At least it will be for me.

For me, I was already bitten by this from the other modules: I assumed that modules like `docker_config` and `docker_secret` is independent of Swarm mode because they didn't have `_swarm_` in them, and after looking a bit into them I was somewhat disappointed that I cannot use them (since I don't use Docker Swarm). But I would have had the same problem if I found out about these functionalities from the docker command line... :)

> So unless @tbouvet disagree with me I opt for naming that align with the docker command line and API.

Fine with me.

Commit:
Creating module_utils/docker_swarm.py for common docker swarm methods to share between modules

Commit:
module_utils/docker_swarm.py: Adding AnsibleDockerSwarmClient class that will extend the AnsibleDockerClient
docker_node: Update documentation section
docker_node: Convert to use the AnsibleDockerSwarmClient for client connection

felixfontein commented:
> You just provided an argument against naming it `docker_swarm_node` instead of `docker_node` - `docker_secret` and `docker_config` should be renamed for consistent approach to `docker_swarm_config` and `docker_swarm_secret`.

I wouldn't mind renaming them as well ;-) But as I said, we can also keep things as they are now (and name the new module `docker_node`).

> Considering the fact the `docker_service` is full of bugs and unmaintained then it might be easier to rename it to `docker_compose` if someone picks it up and have the `docker_swarm_service` renamed to `docker_service`.

Actually, renaming `docker_service` to `docker_compose` is a good idea. That more closely describes what it is. For backwards compatibility, there will be a symlink from the old module name to the new module name for four Ansible versions (that's the standard deprecation cycle). When that's gone, we can think about renaming `docker_swarm_service`.

> If there are different orchestrators in future supported in Docker natively I would like to prepare components like secrets or config in a unified way, just not care about such fact as different orchestrators in playbooks unless I perform orchestrator related actions. Otherwise, it will just increase the complexity of playbooks. Like I said, maybe the Docker developers had the same approach in mind

If/when that happens, it might also make sense to create a orchestrator-agnostic set of modules for managing them, similar to the [package module](https://docs.ansible.com/ansible/latest/modules/package_module.html) which uses `apt`, `yum`, or whatever else is around in the background to install packages. But I guess it's better to wait until that happens and then decide how to proceed, especially since now we don't know how the unified thing will look like ;-)

WojciechowskiPiotr commented:
Moved PR to #50584

tbouvet commented:
@WojciechowskiPiotr I agree with you, I prefer `docker_node` as `docker_swarm_node`. Otherwise, we have to rename `docker_secret` to `docker_swarm_secret`, ...
